### DISCUSSION TITLE
How to update the value of a IState<Uri> using Set/SetAsync

### DISCUSSION THREAD

**Discussion by mcNets** [unoplatform/uno#16724] [Q&A] [answered]
I'm doing some tests with MVUX and I'm facing this error trying to update the value of a IState using Set.

In both cases using a variable or using a new object I get the same error:

> CS0453: The type 'Uri' must be a non-nullable value type in order to use it as a parameter

```csharp
public partial record OsmoViewModel
{
    public IState<Uri> ImgSrc { get; }

    public OsmoViewModel()
    {
        ImgSrc = State<Uri>.Value(this, () => new Uri("ms-appx:///Assets/Images/crono1.jpg"));        
    }

    public async ValueTask ImgCommand(CancellationToken ct = default) 
    {
        Uri newUri = new Uri("ms-appx:///Assets/Images/crono.jpg");
        await ImgSrc.SetAsync(newUri, ct);
        await Task.Delay(2000);
        await ImgSrc.SetAsync(new Uri("ms-appx:///Assets/Images/crono1.jpg"), ct);
    }
}
```

This is the example in the documentation:

```csharp
public async ValueTask SetSliderMiddle(CancellationToken ct = default)
{    
    await SliderValue.Set(50, ct);
}
```

**Comment by francoistanguay** [comment]
Can you use .UpdateAsync( v => ... ) ?

Set has a specific meaning that @dr1rrb  can explain.

**Comment by dr1rrb** [comment]
This is to ensure the ACID properties. The `Set` is only a helper that has been allowed for `ValueType` (i.e. `struct`, `Enum`, `double`, `int`etc.)  and `string` as ACID less relevant for such kind of values. But this is only a shortcut, the right way to update an `IState` is to use the `UpdateAysnc` which accepts the current value as parameter.

**Reply by dr1rrb** [reply, marked as answer]
As this is a recurring friction point, I think we should instead create an analyzer to detect the common issue :

```csharp
var myValue = await MyState;
myUpdatedValue = myValue with { MyProperty = 42 };
MyState.Set(myUpdatedValue);
```

and then allow the `Set` no matter the type of `T`

@francoistanguay any opinion?

> Note: The right way to do example above is
> ```csharp
> MyState.UpdateAsync(myValue => myValue with { MyProperty = 42 });
> ```

**Reply by mcNets** [reply]
Thanks, now it works.

```csharp
    public async ValueTask ImgCommand(CancellationToken ct = default)
    {
        await ImgSrc.UpdateAsync((_) => new Uri("ms-appx:///Assets/Images/crono.jpg"), ct);
        await Task.Delay(2000, ct);
        await ImgSrc.UpdateAsync((_) => new Uri("ms-appx:///Assets/Images/crono2.jpg"), ct);
    }
```
